### ISSUE TITLE
Improve Docker image for efficiency and size

### ISSUE THREAD

**Issue by ibnesayeed** [arXiv/arxiv-base#138]
I suggest the following changes to improve the Dockerfile of this repo:

* [ ] Remove `/tmp/safe_yum.sh` file after it is executed
* [ ] Move environment variable declarations and other commands that do not change often above frequently changing layers for better caching
* [ ] Combine multiple `ENV` declarations in one to minimize layers

**Comment by erickpeirson** [comment]
I'll add to this:

- [ ] Since arxiv-base package requires mysql, we should add ``mysql mysql-devel`` to the yum install step.
- [ ] For images that extend ``arxiv/base`` (e.g. for a particular service), we should start putting pip-related commands on a single step, ending with ``&& rm -rf ~/.cache/pip``. For example:

```docker
RUN pip install -U pip pipenv uwsgi \
    && pipenv install \
    && rm -rf ~/.cache/pip
```

**Comment by ibnesayeed** [comment]
> Since arxiv-base package requires mysql, we should add mysql mysql-devel to the yum install step.

Shouldn't we keep MySQL in a separate container (as a microservice), utilizing the official image?

**Comment by erickpeirson** [comment]
> > Since arxiv-base package requires mysql, we should add mysql mysql-devel to the yum install step.
> 
> Shouldn't we keep MySQL in a separate container (as a microservice), utilizing the official image?

IIRC, the problem is that the ``mysqlclient`` Python package expects at least the MySQL client to be installed locally

**Comment by ibnesayeed** [comment]
> IIRC, the problem is that the `mysqlclient` Python package expects at least the MySQL client to be installed locally

It looks like [the package](https://pypi.org/project/mysqlclient/) requires Python and MySQL development headers and libraries, but not the client itself (unless I missed something).

**Comment by erickpeirson** [comment]
> It looks like [the package](https://pypi.org/project/mysqlclient/) requires Python and MySQL development headers and libraries, but not the client itself (unless I missed something).

Yeah, you're right. So since we're on Centos7, should be just ``mariadb-devel`` I think?

**Comment by ibnesayeed** [comment]
... and `python3-devel` too I guess?

**Comment by erickpeirson** [comment]
Want to take a look at https://github.com/arXiv/arxiv-base/pull/139 and see what you think so far?

**Comment by ibnesayeed** [comment]
Looks like it is taking a good shape. I have posted a couple minor comments/suggestions there.

**Comment by ibnesayeed** [comment]
I have checked the check boxes in the first post as per #139. Since this is a major rework on the base image, Dockerfiles of other derived images in various other repos should be improved/adjusted accordingly and tested well before making it a go!
